A51-xds-custom-backend-metrics.md

```diff
@@ -0,0 +1,382 @@
+And the load reporting server will keep sending the metrics data regardless of whether the data has changed or not.
+There is a potential optimization, that is to skip sending the unchanged metrics data since last response. 
+To do that, we should not compare maps for de-duplication, instead, we can bump up a 
+generation_id of the metrics data every time users make a change. 
+Each client connection only needs to compare its previous generation_id to do deduplication. 
+The optimization strategy can be implemented until there are concrete use cases in the future.
+
+#### The Client API 
+
+Out-of-band data will be sent only if the client starts a streaming call for the `OpenRCAService/StreamCoreMetrics` method.
+If the backend is designed to provide this information but the client never asks for it, it will not be used.  
+Conversely, if the backend does not provide the information but the client does start a call asking for it,
+the backend will terminate the call with status `UNIMPLEMENTED`, in which case the client will not retry the call.
+
+Any LB policy that is interested in this out-of-band data may subscribe to it.  
+The client will maintain an open `OpenRCAService/StreamCoreMetrics` call 
+whenever there is at least one LB policy subscribing to the data.
+
+When a policy subscribes, it will indicate the interval at which it would like to get updates. 
+If multiple policies subscribe with different intervals, we will use the minimum of those intervals.  
+Note that the ORCA protocol involves the client telling the server the interval at which 
+it wants reports when the call starts, so if the client needs to change its requested interval, 
+it will need to cancel and restart the call.
+A client may change its requested interval at any time; this is expected to be a fairly rare operation 
+(usually triggered by a config change for one of the subscribing LB policies),
+It is important that a load balancing policy is able to re-config the frequency
+and should not require heavy-weight operations like reconnecting to backends or recreating subchannels or LB policies.
+
+Each policy that has subscribed to the data will get a callback whenever a new message is sent to the client on the call.  
+As with per-request data, the protobuf deserialization will happen only once.  
+The data provided to the callback should be exposed to the LB policy using the same type as used for per-request data.
+
+The API for subscribing to out-of-band reports must be available to LB policies at any level of the routing hierarchy, 
+including both leaf LB policies (whose children are subchannels) and parent LB policies (whose children are other LB policies).  
+Note that there may be multiple LB policies in the routing hierarchy that are interested in subscribing.  
+For example, in the case where we are using the xDS policy for cluster routing and the WRR policy for intra-cluster routing,
```

Comment:
The wording here is a little out of date; it dates back to when all of the xDS functionality was in a single "xds" LB policy.  I suggest changing it to say something like this:

"""
For example, when using xDS, the `xds_cluster_impl` policy may need to see backend metric data to generate load reports for each cluster, and the WRR policy might need to see the same data for endpoint-picking within a locality.
"""